### PR TITLE
Sample name cleaning with pairs of input filenames

### PR DESCRIPTION
**Pull request by vladsavelyev**
Address https://github.com/ewels/MultiQC/issues/2162

Make `self.clean_s_name` support a lists of paths. Each name will be cleaned separately first, and then a common prefix will be found. If the common prefix is empty, the cleaned names will be dash-concatenated. Since the use case if primarily for pairs of FASTQ files, common FASTQ suffixes will be trimmed.

First check for Illumina conventional [naming](https://support.illumina.com/help/BaseSpace_Sequence_Hub_OLH_009008_2/Source/Informatics/BS/NamingConvention_FASTQ-files-swBS.htm):

```py
>>> self.clean_s_name(["EDM_TAGGCA_L001_R1_001", "EDM_TAGGCA_L001_R2_001"])
"EDM_TAGGCA_L001"
```

Other cases when `_R1`/`_R2` is anywhere:

```py
>>> self.clean_s_name(["sample_1_R1_trimmed", "sample_1_R2_trimmed"])
"sample_1_trimmed"
>>> self.clean_s_name(["TE18-018_12S_R1", "TE18-018_12S_R2"])
"TE18-018_12S"
```

We don't want to remove anything less conventional from the middle, so we only check for other variations in the end (.1, _1, r1, etc):

```py
>>> self.clean_s_name(["fastq_r1", "fastq_r2"])
"fastq"
>>> self.clean_s_name(["Sample_1_1", "Sample_1_2"])
"Sample_1"
>>> self.clean_s_name(["SAMPLE1_PE.1", "SAMPLE1_PE.2")
"SAMPLE1_PE"
```

Adjusting read processing modules to make use of this feature and pass found FASTQ pairs to `self.clean_s_name`:

- [x] trimmomatic
- [x] cutadapt
- [x] fastp

-----

Additionally, fixes a bug in `cutadapt` when an option parameter was taken as a sample name:

```sh
head -n2 issue_1328/SRR4238351_subsamp_m_9999999999.cutadapt.log
This is cutadapt 3.4 with Python 3.9.2
Command line parameters: -a GGGGGGGGGGATATGCGATCGC...GTGTCTCAACGGTATCCGCAACT -o SRR4238351_subsamp.trimed.fastq.gz SRR4238351_subsamp.fastq.gz -m 999999999999999
```
Screenshot 2023-11-16 at 15 10 18

### CODE REVIEW

File: multiqc/modules/base_module.py [hunk 1]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+
+        if len(clean_names) == 2:
+            # Possibly a FASTQ pair - try trimming FASTQ suffixes.
+            r1, r2 = sorted(clean_names)
+
+            # Try trimming the conventional illumina suffix with a tail 001 ending. Refs:
+            # https://support.illumina.com/help/BaseSpace_Sequence_Hub_OLH_009008_2/Source/Informatics/BS/NamingConvention_FASTQ-files-swBS.htm
+            # https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/using/fastq-input#:~:text=10x%20pipelines%20need%20files%20named,individual%20who%20demultiplexed%20your%20flowcell.
+            cleaned_r1 = re.sub(r"_R1_\d\d\d$", "", r1)
+            cleaned_r2 = re.sub(r"_R2_\d\d\d$", "", r2)
```
Comment (ewels) [inline multiqc/modules/base_module.py:300]:
```suggestion
            cleaned_r1 = re.sub(r"_R1_\d{3}$", "", r1)
            cleaned_r2 = re.sub(r"_R2_\d{3}$", "", r2)
```

File: multiqc/modules/base_module.py [hunk 2]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+            if cleaned_r1 == cleaned_r2:  # trimmed successfully
+                return cleaned_r1
```
Comment (ewels) [inline multiqc/modules/base_module.py:302]: If we don't return here, we continue with partially-cleaned sample names, right? Do we want to revert to the full filenames before continuing on to the next attempt?

File: multiqc/modules/base_module.py [hunk 3]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+
+            # Try removing _R1 and _R2 from the middle.
+            cleaned_r1 = re.sub(r"_R1_", "_", r1)
+            cleaned_r2 = re.sub(r"_R2_", "_", r2)
+            if cleaned_r1 == cleaned_r2:  # trimmed successfully
+                return cleaned_r1
+
+            # Try trimming other variations from the end (-R1, _r1, _1, .1, etc).
+            cleaned_r1 = re.sub(r"([_.-][rR]?1)?$", "", r1)
+            cleaned_r2 = re.sub(r"([_.-][rR]?2)?$", "", r2)
```
Comment (ewels) [inline multiqc/modules/base_module.py:312]:
```suggestion
            cleaned_r1 = re.sub(r"([_\.-][rR]?1)?$", "", r1)
            cleaned_r2 = re.sub(r"([_\.-][rR]?2)?$", "", r2)
```

File: multiqc/modules/base_module.py [hunk 4]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+            if cleaned_r1 == cleaned_r2:  # trimmed successfully
+                return cleaned_r1
+
+        # More than 2 names, so can't assume a FASTQ pair. Just trying to keep a common prefix.
```
Comment (ewels) [inline multiqc/modules/base_module.py:316]: Not necessarily more than 2 names, could be 2 but weren't able to trim.

```suggestion
        # Couldn't clean, or more than 2 names. Just trying to keep a common prefix.
```

File: multiqc/modules/base_module.py [hunk 5]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+            if len(s_name) == 0:
+                raise ValueError("Empty list of sample names passed to clean_s_name()")
+            if len(s_name) >= 2:
+                # Extract a sample name from a list of file names (for example, FASTQ pairs).
+                # Each name is cleaned separately first.
+                clean_names = [
+                    self.clean_s_name(sn, f=f, root=root, filename=filename, seach_pattern_key=seach_pattern_key)
+                    for sn in s_name
+                ]
+                return self._clean_fastq_name_pair(clean_names)
```
Comment (ewels) [inline multiqc/modules/base_module.py:344]: I think that we don't want to return yet, no?

eg. if two identical filenames `sample.fastq.gz` and `sample.fastq.gz` are given, it'll return the full `sample.fastq.gz` instead of cleaning to `sample`.

Also, returning at this stage means we miss directory prefixes etc.

File: multiqc/modules/base_module.py [hunk 6]
```diff
@@ -281,14 +281,70 @@ def add_section(
...
+                return cleaned_r1
+
+        # More than 2 names, so can't assume a FASTQ pair. Just trying to keep a common prefix.
+        prefix = os.path.commonprefix(clean_names)
+        prefix = prefix.rstrip("_.- ")
+        # If the prefix is empty, join the names with a dash.
+        if prefix == "":
+            s_name = "_".join(clean_names)
+        else:
+            s_name = prefix
```
Comment (ewels) [inline multiqc/modules/base_module.py:323]: Not sure if this is too generic 🤔  Need to think about some edge cases to test it with.

**Review by ewels — COMMENTED** [review]
Nice, couple of questions / comments but basically good to go I think.

File: multiqc/modules/base_module.py (hunk 1 above)
Comment (vladsavelyev) [reply multiqc/modules/base_module.py:300]: I personally prefer to duplicate, but that's alright :)

File: multiqc/modules/base_module.py (hunk 3 above)
Comment (vladsavelyev) [reply multiqc/modules/base_module.py:312]: Not needed inside square brackets

File: multiqc/modules/base_module.py (hunk 5 above)
Comment (vladsavelyev) [reply multiqc/modules/base_module.py:344]: > eg. if two identical filenames `sample.fastq.gz` and `sample.fastq.gz` are given, it'll return the full `sample.fastq.gz` instead of cleaning to `sample`.

It will be cleaned to two identical [`sample`, `sample`] in the list comprehension above. That's why the variable is called `clean_names` :)

File: multiqc/modules/base_module.py (hunk 2 above)
Comment (vladsavelyev) [reply multiqc/modules/base_module.py:302]: We clean the full names `r1`/`r2` on every attempt, and `cleaned_r1`/`cleaned_r2` are discarded

Comment (ewels) [reply multiqc/modules/base_module.py:302]: Ah you're right, was misreading sorry 👍🏻

File: multiqc/modules/base_module.py (hunk 3 above)
Comment (ewels) [reply multiqc/modules/base_module.py:312]: huh, nice!

File: multiqc/modules/base_module.py (hunk 5 above)
Comment (ewels) [reply multiqc/modules/base_module.py:344]: Ooh I didn't notice that - calling the function on itself, nice!

File: multiqc/modules/base_module.py (hunk 6 above)
Comment (ewels) [reply multiqc/modules/base_module.py:323]: So am I right in thinking that the following would be true here?

`sample_1` + `sequence_2` goes to `s`

Comment (vladsavelyev) [reply multiqc/modules/base_module.py:323]: Yeah, that's right. Not sure how common that usecase is. Perhaps we should always "_"-merge?

Comment (vladsavelyev) [reply multiqc/modules/base_module.py]: Done!

**Review by ewels — APPROVED** [review]
Code LGTM 👍🏻
